File: ios/SharedList/Core/Auth/AuthDTOs.swift
import Foundation

// Wire types that mirror the backend's auth schemas (backend/src/features/auth/schemas.ts).
//
// We keep these structs Codable + Sendable + Equatable. Sendable is the
// important one for Swift 6: any value crossing an actor boundary (an API
// call response decoded on the URLSession actor, then handed to TokenStore
// on the MainActor) needs Sendable. Plain value types whose stored properties
// are all Sendable get the conformance implicitly, but stating it explicitly
// is a useful sanity check — if a non-Sendable property sneaks in later,
// the compiler points at this declaration.
//
// Naming mirrors the backend types verbatim so a `git grep AuthResponse`
// finds matches across the whole repo. JSON keys also match the wire format
// (camelCase) — no Swift-side renaming dance.

public struct AuthUser: Codable, Sendable, Equatable, Hashable {
    public let id: String
    public let email: String
    public let displayName: String

    public init(id: String, email: String, displayName: String) {
        self.id = id
        self.email = email
        self.displayName = displayName
    }
}

public struct AuthResponse: Codable, Sendable, Equatable {
    public let user: AuthUser
    public let accessToken: String
    public let refreshToken: String
}

public struct SignupBody: Codable, Sendable, Equatable {
    public let email: String
    public let password: String
    public let displayName: String

    public init(email: String, password: String, displayName: String) {
        self.email = email
        self.password = password
        self.displayName = displayName
    }
}

public struct LoginBody: Codable, Sendable, Equatable {
    public let email: String
    public let password: String

    public init(email: String, password: String) {
        self.email = email
        self.password = password
    }
}

public struct RefreshBody: Codable, Sendable, Equatable {
    public let refreshToken: String
}

public struct LogoutBody: Codable, Sendable, Equatable {
    public let refreshToken: String
}

// Error envelope from the backend (backend/src/infra/middleware/error.ts).
// Every non-2xx response carries this shape; APIClient surfaces the inner
// `code` + `requestId` to the caller via APIError so a "weird thing happened"
// bug report can be matched to a server-side log line.
public struct APIErrorEnvelope: Codable, Sendable, Equatable {
    public struct Inner: Codable, Sendable, Equatable {
        public let code: String
        public let message: String
        public let requestId: String
    }
    public let error: Inner
}

File: android/app/src/main/java/in/santosh_bharadwaj/sharedlist/core/auth/AuthDtos.kt
package `in`.santosh_bharadwaj.sharedlist.core.auth

import kotlinx.serialization.Serializable

/**
 * Wire types that mirror the backend's auth schemas (backend/src/features/auth/schemas.ts).
 *
 * @Serializable lets the kotlinx.serialization plugin generate JSON
 * encoders/decoders at compile time — closest cultural mirror to Swift's
 * Codable. JSON keys are camelCase to match the wire format and the iOS DTOs
 * verbatim, so a `git grep AuthResponse` finds matches across all three
 * platforms.
 *
 * Naming mirrors backend types (and iOS DTOs) exactly: `<Verb><Subject>Body`
 * for requests, `<Subject>Response` for responses.
 */
@Serializable
public data class AuthUser(
    public val id: String,
    public val email: String,
    public val displayName: String,
)

@Serializable
public data class AuthResponse(
    public val user: AuthUser,
    public val accessToken: String,
    public val refreshToken: String,
)

@Serializable
public data class SignupBody(
    public val email: String,
    public val password: String,
    public val displayName: String,
)

@Serializable
public data class LoginBody(
    public val email: String,
    public val password: String,
)

@Serializable
public data class RefreshBody(
    public val refreshToken: String,
)

@Serializable
public data class LogoutBody(
    public val refreshToken: String,
)

/**
 * Error envelope from the backend (backend/src/infra/middleware/error.ts).
 * Every non-2xx response carries this shape; ApiClient surfaces the inner
 * `code` + `requestId` to callers via [ApiError.Server].
 */
@Serializable
public data class ApiErrorEnvelope(
    public val error: ApiErrorBody,
)

@Serializable
public data class ApiErrorBody(
    public val code: String,
    public val message: String,
    public val requestId: String,
)
